rclcpp/test/rclcpp/strategies/test_message_pool_memory_strategy.cpp

```diff
@@ -0,0 +1,64 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+
+#include "gtest/gtest.h"
+
+#include "rclcpp/strategies/message_pool_memory_strategy.hpp"
+#include "test_msgs/msg/empty.hpp"
+
+using rclcpp::strategies::message_pool_memory_strategy::MessagePoolMemoryStrategy;
+
+class TestMessagePoolMemoryStrategy : public ::testing::Test
+{
+public:
+  void SetUp()
+  {
+    message_memory_strategy_ =
+      std::make_shared<MessagePoolMemoryStrategy<test_msgs::msg::Empty, 1>>();
+  }
+
+protected:
+  std::shared_ptr<MessagePoolMemoryStrategy<test_msgs::msg::Empty, 1>> message_memory_strategy_;
+};
+
+TEST_F(TestMessagePoolMemoryStrategy, construct_destruct) {
+}
+
+TEST_F(TestMessagePoolMemoryStrategy, borrow_return) {
+  auto message = message_memory_strategy_->borrow_message();
+  ASSERT_NE(nullptr, message);
+
+  EXPECT_NO_THROW(message_memory_strategy_->return_message(message));
+}
+
+TEST_F(TestMessagePoolMemoryStrategy, borrow_too_many) {
+  auto message = message_memory_strategy_->borrow_message();
+  ASSERT_NE(nullptr, message);
+
+  // Size is 1, borrowing second time should fail
+  EXPECT_THROW(message_memory_strategy_->borrow_message(), std::runtime_error);
```

Comment:
I sometimes find that these checks are too generic.  In particular, there are lots of things in the system that can throw `std::runtime_error`, and there are several exception classes that derive from it.  So you are not always sure which exception you are getting.  In the past, I've used macros like this:

```
#define EXPECT_THROW_OF_TYPE(type, statement, msg)                                                           \
  do {                                                                                                       \
    try {                                                                                                    \
      statement;                                                                                             \
    } catch (type const& err) {                                                                              \
      if (std::string(err.what()).find(msg) == std::string::npos) {                                          \
        FAIL() << "Expected error msg containing:" << std::endl                                              \
               << msg << std::endl                                                                           \
               << "Saw error msg:" << std::endl                                                              \
               << err.what() << std::endl;                                                                   \
      }                                                                                                      \
    } catch (std::exception const& err) {                                                                    \
      FAIL() << "Expected " #type << std::endl << "Saw exception type: " << typeid(err).name() << std::endl; \
    }                                                                                                        \
  } while (0)

#define EXPECT_RUNTIME_THROW(st, msg) EXPECT_THROW_OF_TYPE(std::runtime_error, st, msg)
#define EXPECT_ARGUMENT_THROW(st, msg) EXPECT_THROW_OF_TYPE(std::invalid_argument, st, msg)
```

It's unfortunately quite wordy, but the benefit is that it tests that the exception is what you expect *and* that the error string is also what you expect.  What do you think about using something like that?

Reply:
Finally got back to this. See PR #1232 for the added macros. I took a slightly different tact, what do you think?

Reply:
Overall, looking good.  I left some comments on #1232 , so resolving here.